Sources/Run/Services/ProductPaymentHandler.swift

```diff
@@ -1,8 +1,108 @@
+        guard let stripe_customer_id = subscriber.stripe_customer_id else {
+            analytics?.logError("Can't process purchase for existing client with missing stripe_customer_id")
+            self.send(message: "Something went wrong. Please request your horoscope again.",
+                      senderId: subscriber.fb_messenger_id,
+                      messagingType: .RESPONSE)
+            return
+        }
+        
+        let (_, price) = self.priceForCode(code)
+        processPaymentWithCardOnFile(subscriber: subscriber, price: price, product: product, stripeCustomerId: stripe_customer_id)
+    }
+    
+    func processPaymentWithCardOnFile(subscriber: Subscriber, price: Double, product: String, stripeCustomerId: String) {
+        if let response = stripeClient?.stripeCharge(subscriber: subscriber,
+                                                     description: product,
+                                                     amount: Int(price * 100),
+                                                     stripeCustomerId: stripeCustomerId) {
+            if response.status == .ok {
+                DispatchQueue.global(qos: DispatchQoS.QoSClass.default).async {
+                    analytics?.logAnalytics(event: .CompletedPurchase,
+                                            for: subscriber,
+                                            eventValue: stripeCustomerId)
+                    //get a spot in the show
+                }
+            } else {
+                handleStripeClientChargeFailure(subscriber: subscriber,
+                                                stripe_customer_id: stripeCustomerId)
+            }
+        } else {
+            handleStripeClientChargeFailure(subscriber: subscriber,
+                                            stripe_customer_id: stripeCustomerId)
+        }
+        
+    }
+    
+    func handleStripeClientChargeFailure(subscriber: Subscriber, stripe_customer_id: String) {
+        analytics?.logError("Failed to charge \(subscriber), stripe_client_id=\(stripe_customer_id)")
+        
+        DispatchQueue.global(qos: DispatchQoS.QoSClass.default).async {
+            subscriber.stripe_customer_id = nil
+            subscriber.remember_card_on_file = false
+        }
+    }
+    
+    func handleMessengerPurchase(subscriber: Subscriber,
+                                 token: String,
+                                 product: String,
+                                 host: String,
+                                 price: String,
+                                 email: String?,
+                                 remember: Bool) throws -> Response {
+        if let result = stripeClient?.processChargeFor(subscriber: subscriber,
+                                                       token: token,
+                                                       description: "charge for \(product)",
+                                                       amount: Int(price.int! * 100),
```

Comment:
p0: Please don't use force unwrap. If value will be nil, app will be crashed. This is vary bad behavior use guard or default value instead.